### ISSUE TITLE
No metrics-values for nomad.client.allocs.memory.cache and wrong values of resource utilization in the nomad-ui after upgrading to Ubuntu 24.04

### ISSUE THREAD

**Issue by SCORP111** [hashicorp/nomad#24524]
### Nomad version
1.9.3

### Operating system and Environment details
Ubuntu 24.0.4

### Issue
After upgrading from version 1.8.1 to 1.9.3 we are seeing weird memory consumption values of our tasks. The tasks keep rising to almost 100% memory consumption.

Example of one task:

![grafik](https://github.com/user-attachments/assets/439b6f75-d8ac-4e37-9996-24c064c07dc6)
![grafik](https://github.com/user-attachments/assets/a182a52e-7049-47a6-98ad-e8609df4e055)

Also in the Nomad-UI the resource utilization is displayed incorrectly:
![grafik](https://github.com/user-attachments/assets/d2e33035-91f3-4a05-9c32-5198cb1ff02e)

### Reproduction steps
Upgrade to 1.9.3
#### Expected Result
Memory consumption is stable as before.
#### Actual Result
Memory consumption keeps rising

**Comment by pkazmierczak** [comment]
Hey @SCORP111, thanks for reporting this issue. Do your tasks all use docker driver by any chance? I'm asking because we've recently made changes to the docker driver, and I wonder if that could be related.

**Comment by SCORP111** [comment]
> Hey @SCORP111, thanks for reporting this issue. Do your tasks all use docker driver by any chance? I'm asking because we've recently made changes to the docker driver, and I wonder if that could be related.

Yes, apart from the pre-start-task, which is using raw_exec, all tasks are using the docker driver. Thanks for your help!

**Comment by pkazmierczak** [comment]
@SCORP111 can you perhaps provide some more details about your jobs?

**Comment by SCORP111** [comment]
This is an example jobfile where the behavious happens:

```
job "[[.service.name]]-[[.environment.stage]]" {
  meta {
    time = "[[ timeNow ]]"
    version = "[[.version]]"
  }
  datacenters = "[[.environment.datacenter]]"
  namespace = "[[.job.namespace]]"
  type = "service"

  reschedule {
    delay = "30s"
    delay_function = "constant"
    unlimited = true
  }

  update {
    max_parallel     = 1
    min_healthy_time = "20s"
    healthy_deadline = "20m"
    progress_deadline = "30m"
  }

  group "testservice" {
    count = [[.group.count]]
    constraint {
      operator  = "distinct_hosts"
      value     = "true"
    }
    network {
      mode = "bridge"
      port "http" {}
      hostname = "${node.unique.name}"
    }

    task "testservice" {
      driver = "docker"
      resources {
        memory = [[.resources.nomadmemory]]
      }
      config {
        image = "[[.config.registry]]/services/testentication-service:[[.version]]"
        ports = ["http"]
      }
      env {
        ENV_JVM_ARGS = "-Djava.awt.headless=true -server -Xms64m -Xmx[[.resources.jvmmemory]]m -XX:+DisableExplicitGC -XX:+UseG1GC -XX:MaxGCPauseMillis=50 -XX:+UseStringDeduplication -XX:+PrintFlagsFinal -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=${NOMAD_ALLOC_DIR}/data -Xlog:gc*,gc+age=trace,gc+promotion=debug,gc+ergo*=trace:file=${NOMAD_ALLOC_DIR}/data/gc.log:time,level,tags:filecount=4,filesize=100M"
        ENV_APP_ARGS = "-Dspring.config.additional-location=file:/secrets/config/testservice-config.yml"
        TZ = "Europe/Berlin"
      }
      artifact {
        source = "git::ssh://git@git.test.info//config//[[.environment.stage]]"
        destination = "local/config"
        options {
          sshkey = "[[ fileContents "/data/id_rsa" | b64enc ]]"
        }
      }
      artifact {
        source = "git::ssh://git@git.test.info//certificates//[[.environment.stage]]"
        destination = "local/certificates"
        options {
          sshkey = "[[ fileContents "/data/id_rsa" | b64enc ]]"
        }
      }
      artifact {
        source = "git::ssh://git@git.test.info//logging//[[.environment.stage]]"
        destination = "local/logging"
        options {
          sshkey = "[[ fileContents "/data/id_rsa" | b64enc ]]"
        }
      }
      template {
        source = "local/config/testservice-config.tmpl"
        destination = "secrets/config/testservice-config.yml"
        # Specifies the behavior Nomad should take if the rendered template changes. Nomad will always write the new contents of the template to the specified destination.
        change_mode = "noop"
      }
      template {
        data = <<EOH
          SERVICEAPIB64="{{key "secrets/testservice_base64"}}"
        EOH

        destination = "secrets/config.env"
        env         = true
        change_mode = "noop"
      }
      service {
        name = "[[.service.name]]-[[.environment.stage]]"
        port = "http"
        check {
          name = "healthcheck_testservice"
          type     = "http"
          tls_skip_verify = true
          protocol = "https"
          path = "/serviceApi/actuator/health"
          interval = "15s"
          timeout  = "15s"
          header {
            Authorization = ["Basic ${SERVICEAPIB64}"]
          }
        }
      }
    }
  }
}
```

I also downgraded my clients to 1.8.1 with Ubuntu 20.04. Now the Nomad-UI shows the correct values again. I will observe the memory behaviour over the weekend now and do some additional tests on monday. Maybe its not caused by the nomad-upgrade, but the different Ubuntu and/or docker-version.

**Comment by SCORP111** [comment]
Testresults:
* Normal behaviour when downgrading to 1.8.1 again, as soon as I upgrade the clients to 1.9.3 the memory consumption keeps rising again
* When doing the initial upgrade, we also upgraded from Ubuntu 20.04 to 24.04 and Docker 27.0.3 to 27.3.1. I also reverted these changes and the memory problem also appears with Ubuntu 20.04 and Docker 27.0.3

![grafik](https://github.com/user-attachments/assets/6f46079a-404b-4f26-bce5-d943d112e187)

When running docker stats for a task I get these values:
```
CONTAINER ID   NAME                                                           CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O        PIDS
d324403f8aac   ...-30039263-c2f2-c09f-4936-4f325af0d180                       4.98%     603.5MiB / 1.25GiB    47.15%    137MB / 477MB     0B / 0B          185
```

While the UI shows these values:
![grafik](https://github.com/user-attachments/assets/7cd1f0e6-d402-4b77-87b0-7acb22fefc90)

### Assumption
Something changed for providing the metric **nomad_client_allocs_memory_cache** (maybe this is related? https://github.com/hashicorp/nomad/issues/24339):
![grafik](https://github.com/user-attachments/assets/cbcdb15f-e0de-4fa4-9be4-5278dd15523e)

As for the 2nd problem: Wrong values in the UI seem to be displayed when using raw_exec as a prestart-task:
![grafik](https://github.com/user-attachments/assets/498f052f-ad8c-48a5-9117-87ad37a1d4e4)

**Comment by SCORP111** [comment]
The incorrect display of values in the nomad-ui only seems to occur with our tasks that are using raw_exec as prestart_tasks I just noticed.

**Comment by SCORP111** [comment]
Alright this is what I found out now:

Downgraded 4 nomad-clients from cgroupv2 to cgroupv1, still using Nomad 1.8.1, metrics for `nomad.client.allocs.memory.cache ` have the correct values again:
![grafik](https://github.com/user-attachments/assets/0c262d48-5597-4960-9032-cabb78877453)
Nomad-UI also working again:
![grafik](https://github.com/user-attachments/assets/c39f6f4c-14d9-46c6-bbe0-9bb733359fa7)

I then downgraded 4 clients in another cluster from cgroupv2 to cgroupv1, these clients were running Nomad 1.9.3, but nothing changed. Only after I downgraded one client to 1.8.1 metrics had the correct values again and the nomad-ui also worked. After upgrading to 1.9.3 again, the metrics had no value again:
![grafik](https://github.com/user-attachments/assets/a6d8d44e-c7e0-444c-b52c-92ab08f0ee0f)

Seems to be a weird bug(?) with cgroup- and/or nomad-version?

**Comment by jedd** [comment]
Hey @SCORP111  -- if you have the capability, can you please try some version checks in between the 1.8.1 and 1.9.3 range?

Refer:  https://github.com/hashicorp/nomad/issues/24339#issuecomment-2466184369

Specifically there's another issue that's been raised that happened in the **1.9.0 -> 1.9.1** change, and the actual commit (out of the dozen or so in between those minor rel's) that looks to be the culprit.

Compiling a custom nomad binary isn't too messy, but would be great to have that extra datapoint.

**Comment by SCORP111** [comment]
> Hey @SCORP111 -- if you have the capability, can you please try some version checks in between the 1.8.1 and 1.9.3 range?
> 
> Refer: [#24339 (comment)](https://github.com/hashicorp/nomad/issues/24339#issuecomment-2466184369)
> 
> Specifically there's another issue that's been raised that happened in the **1.9.0 -> 1.9.1** change, and the actual commit (out of the dozen or so in between those minor rel's) that looks to be the culprit.
> 
> Compiling a custom nomad binary isn't too messy, but would be great to have that extra datapoint.

I downgraded a client Using Ubuntu 24.04 (with cgroupv1) from 1.9.3 to 1.9.0, but metrics where not emitted. I then downgraded to 1.8.4 and metrics were emitted again. Let me know if I can help with any other tests. Thank you!

**Comment by pkazmierczak** [comment]
Hey @SCORP111, we've been making some changes to telemetry and docker metrics especially. Upcoming 1.9.5 release will contain our most recent fixes, i.e.: https://github.com/hashicorp/nomad/pull/24768

I will close this issue for now, but feel free to re-open or create a new one if the problem persists on 1.9.5.
